integer.RoundToInt32 fails for 0.49999999999999994

If you add the following test to `integer/integer_test.go`:

```
{
  num: 0.49999999999999994,
  exp: 0,
}
```

And then run the tests, they fail:

```
go test ./integer
--- FAIL: TestRoundToInt32 (0.00s)
	integer_test.go:243: executing scenario 0
	integer_test.go:243: executing scenario 1
	integer_test.go:243: executing scenario 2
	integer_test.go:243: executing scenario 3
	integer_test.go:243: executing scenario 4
	integer_test.go:243: executing scenario 5
	integer_test.go:245: expected 0, got 1
	integer_test.go:243: executing scenario 6
FAIL
FAIL	github.com/kubernetes/utils/integer	0.006s
```

Is this beyond the scope of this function? If not I can submit a PR to fix this, though would it be acceptable to use `math.Round` provided in Go 1.10 and above?

If math.Round now offers the same or better behavior, it's fine to switch
to that.

On Thu, Jan 24, 2019 at 7:43 PM Matias Anaya <notifications@github.com>
wrote:

> If you add the following test to integer/integer_test.go:
>
> {
>   num: 0.49999999999999994,
>   exp: 0,
> }
>
> And then run the tests, they fail:
>
> go test ./integer
> --- FAIL: TestRoundToInt32 (0.00s)
> 	integer_test.go:243: executing scenario 0
> 	integer_test.go:243: executing scenario 1
> 	integer_test.go:243: executing scenario 2
> 	integer_test.go:243: executing scenario 3
> 	integer_test.go:243: executing scenario 4
> 	integer_test.go:243: executing scenario 5
> 	integer_test.go:245: expected 0, got 1
> 	integer_test.go:243: executing scenario 6
> FAIL
> FAIL	github.com/kubernetes/utils/integer	0.006s
>
> Is this beyond the scope of this function? If not I can submit a PR to fix
> this, though would it be acceptable to use math.Round provided in Go 1.10
> and above?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/utils/issues/71>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVMrhxR3hgaUxMC9x2PISj7SZ-lLYks5vGn1RgaJpZM4aSJmh>
> .
>

--- a/integer/integer.go
+++ b/integer/integer.go
@@ -16,6 +16,8 @@ limitations under the License.
 
 package integer
 
+import "math"
+
 // IntMax returns the maximum of the params
 func IntMax(a, b int) int {
 	if b > a {
@@ -65,9 +67,7 @@ func Int64Min(a, b int64) int64 {
 }
 
 // RoundToInt32 rounds floats into integer numbers.
+// Deprecated: use math.Round() and a cast directly.
 func RoundToInt32(a float64) int32 {
-	if a < 0 {
-		return int32(a - 0.5)
-	}
-	return int32(a + 0.5)
+	return int32(math.Round(a))
 }

--- a/integer/integer_test.go
+++ b/integer/integer_test.go
@@ -233,6 +233,10 @@ func TestRoundToInt32(t *testing.T) {
 			num: 0,
 			exp: 0,
 		},
+		{
+			num: 0.49999999999999994,
+			exp: 0,
+		},
 	}
 
 	for i, test := range tests {

